src/polyline2.cpp

```diff
@@ -89,33 +89,75 @@ Point2 Polyline2::PointAt(float route_pos, bool* is_vertex,
                  remaining > segments_[idx].Length();
   }
 
-  return segments_[idx].FirstPoint() +
+  Point2 return_point = segments_[idx].FirstPoint() +
          remaining * segments_[idx].UnitDirection();
+  if (is_endpoint) {
+    if(idx == 0){
+      if(return_point == segments_[idx].FirstPoint())
+        *is_endpoint = true;
+    }
+    else if(idx == segments_.size()-1){
+      if(return_point == segments_[idx].SecondPoint())
+        *is_endpoint = true;
+    }
+    else
+      *is_endpoint = false;
+  }
+
+  return return_point;
 }
 
 Point2 Polyline2::ClosestPoint(const Point2& query, bool* is_vertex,
                                LineSegment2* segment,
-                               float* signed_squared_distance) const {
+                               float* signed_squared_distance, bool* is_endpoint) const {
   // Walk along each line segment and remember which was closest.
   float closest_signed_squared_distance = constants::kInfinity;
   Point2 closest_point;
 
   float current_signed_squared_distance;
+  int segment_ind = 0;
+  int counter = 0;
   for (const auto& s : segments_) {
-    bool is_endpoint;
+    bool is_segment_endpoint;
     const Point2 current_point =
-        s.ClosestPoint(query, &is_endpoint, &current_signed_squared_distance);
+        s.ClosestPoint(query, &is_segment_endpoint, &current_signed_squared_distance);
 
     if (std::abs(current_signed_squared_distance) <
         std::abs(closest_signed_squared_distance)) {
       closest_signed_squared_distance = current_signed_squared_distance;
       closest_point = current_point;
 
-      if (is_vertex) *is_vertex = is_endpoint;
+      if (is_vertex) *is_vertex = is_segment_endpoint;
       if (segment) *segment = s;
+      segment_ind = counter;
     }
+    counter += 1; 
   }
 
+  // check if the closest point occurs at an endpoint for the polyline
+  if (is_endpoint){
+    if(segment_ind == 0){
+      bool check_vertex;
+      float ssd;
+      segments_[segment_ind + 1].ClosestPoint(query, &check_vertex, &ssd);
```

Comment:
I think this is a little complicated logically, so maybe explain in a comment.

Reply:
done!

Reply:
Nice! Also, can you please get rid of the `float ssd` since the `ClosestPoint()` function just defaults that argument to `nullptr` anyway and you don't use the value here.